common/src/main/java/com/wynntils/models/items/encoding/DataTransformerRegistry.java

```diff
@@ -0,0 +1,196 @@
+    }
+
+    public ErrorOr<EncodedByteBuffer> encodeData(ItemTransformingVersion version, List<ItemData> data) {
+        List<UnsignedByte> bytes = new ArrayList<>();
+
+        for (ItemData itemData : data) {
+            try {
+                ErrorOr<UnsignedByte[]> errorOrEncodedData = encodeData(version, itemData);
+                if (errorOrEncodedData.hasError()) {
+                    return ErrorOr.error(errorOrEncodedData.getError());
+                }
+
+                bytes.addAll(Arrays.asList(errorOrEncodedData.getValue()));
+            } catch (Exception e) {
+                WynntilsMod.error(
+                        "Failed to encode data class " + itemData.getClass().getSimpleName() + "!", e);
+                return ErrorOr.error(
+                        "Failed to encode data class " + itemData.getClass().getSimpleName() + "!");
+            }
+        }
+
+        return ErrorOr.of(EncodedByteBuffer.fromBytes(bytes.toArray(new UnsignedByte[0])));
+    }
+
+    public ErrorOr<List<ItemData>> decodeData(EncodedByteBuffer encodedByteBuffer) {
+        ArrayReader<UnsignedByte> byteReader = encodedByteBuffer.getReader();
+
+        // Handle start data specially
+        ErrorOr<StartData> errorOrStartData = StartDataTransformer.decodeData(byteReader);
+        if (errorOrStartData.hasError()) {
+            return ErrorOr.error(errorOrStartData.getError());
+        }
+
+        return decodeData(errorOrStartData.getValue().version(), byteReader);
+    }
+
+    private ErrorOr<UnsignedByte[]> encodeData(ItemTransformingVersion version, ItemData data) {
+        DataTransformer<ItemData> dataTransformer = (DataTransformer<ItemData>) dataTransformers.get(data.getClass());
+        if (dataTransformer == null) {
+            return ErrorOr.error(
+                    "No data transformer found for " + data.getClass().getSimpleName());
+        }
+
+        return dataTransformer.encode(version, data);
+    }
+
+    private ErrorOr<List<ItemData>> decodeData(ItemTransformingVersion version, ArrayReader<UnsignedByte> byteReader) {
+        List<ItemData> dataList = new ArrayList<>();
+
+        while (byteReader.hasRemaining()) {
+            UnsignedByte dataBlockId = byteReader.read();
+
+            try {
+                DataTransformer<ItemData> dataTransformer = dataTransformers.get(dataBlockId.toByte());
+
+                if (dataTransformer == null) {
+                    return ErrorOr.error("No data transformer found for id " + dataBlockId.value());
+                }
+
+                ErrorOr<ItemData> errorOrData = dataTransformer.decodeData(version, byteReader);
+
+                if (errorOrData.hasError()) {
+                    return ErrorOr.error(errorOrData.getError());
+                }
+
+                dataList.add(errorOrData.getValue());
+            } catch (Exception e) {
+                WynntilsMod.error("Failed to decode data block with id " + dataBlockId.value() + "!", e);
```

Comment:
... and here

Reply:
I see this is a common pattern. Maybe introduce a `ErrorOr.errorAndLog()`, `ErrorOr.loggedError()` or something like that?

Reply:
I think it's rather weird to call it like that. Maybe perhaps add a method like `errorInstance.log()`? Or `WynntilsMod#error(ErrorOr<>)`?

Reply:
I don't really know. I reacted to the duplication of the string. Either you can like create a `String errorMsg = "...."` and then call the log and ErrorOr with the string. Or you can create a common helper that basically does that. Whatever you want to call that, I don't really have any good suggestions.